### PR TITLE
Add pipeline version support to frontend

### PR DESCRIPTION
**Pull request by jingzhang36**
1. NewPipelineVersion page for uploading pipeline and version (replacing old upload pipeline dialog)
2. PipelineVersionList page for listing pipeline versions, e.g., when pipeline gets expanded in our home page
3. RunList page shows version reference as well as pipeline reference. In particular, old runs (created before version is introduced) show pipeline reference, while new runs show version reference.
4. Create run from versions instead of pipelines.
5. Delete version button
6. Clone run when run refers to version or refers to pipeline

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2667)

### CODE REVIEW

**Comment by Bobgy** [comment]
Test failed with
```
Checking formatting...
src/lib/Buttons.ts
src/lib/RunUtils.ts
src/pages/RunList.test.tsx
src/pages/RunList.tsx
Code style issues found in the above file(s). Forgot to run Prettier?
```

File: frontend/tslint.json [hunk 1]
```diff
@@ -17,6 +17,7 @@
     "format": "verbose"
   },
   "rules": {
+    "no-console": false,
```
Comment (Bobgy) [inline frontend/tslint.json]: I think we should keep this rule

File: frontend/src/pages/RunList.tsx [hunk 2]
```diff
@@ -395,9 +398,27 @@ class RunList extends React.PureComponent<RunListProps, RunListState> {
...
+    const pipelineVersionId = RunUtils.getPipelineVersionId(displayRun.run);
     const pipelineId = RunUtils.getPipelineId(displayRun.run);
-    if (pipelineId) {
+    if (pipelineVersionId) {
+      try {
+        const pipelineVersion = await Apis.pipelineServiceApi.getPipelineVersion(pipelineVersionId);
+        const pipelineVersionName = pipelineVersion.name || '';
+        displayRun.pipelineVersion = {
+          displayName: pipelineVersionName,
+          pipelineId: pipelineVersion && pipelineVersion.resource_references && pipelineVersion.resource_references.find(ref => ref.key && ref.key.type && ref.key.type === ApiResourceType.PIPELINE)!.key!.id!,
```
Comment (Bobgy) [inline frontend/src/pages/RunList.tsx]: Extract these to a separate util file makes this a lot easier to read.
For example, you can reference: https://github.com/kubeflow/pipelines/blob/5645f25b684d69dc4c997cb010594b094ae24d32/frontend/src/lib/RunUtils.ts#L77

Note that it also handles the case when network data is not valid and returns an `null` gracefully

File: frontend/src/pages/PipelineDetails.tsx [hunk 3]
```diff
@@ -213,10 +242,39 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+                                Version source
+                              </a>
+                            </div>
+                          </React.Fragment>
+                        )}
                         <div className={css.summaryKey}>Uploaded on</div>
                         <div>{formatDateString(pipeline.created_at)}</div>
                         <div className={css.summaryKey}>Description</div>
-                        <Description description={pipeline.description || ''} />
+                        <div>{pipeline.description}</div>
```
Comment (Bobgy) [inline frontend/src/pages/PipelineDetails.tsx]: You should revert the change here.

File: frontend/src/lib/RunUtils.ts [hunk 4]
```diff
@@ -66,6 +60,18 @@ function getPipelineName(run?: ApiRun | ApiJob): string | null {
...
 }
 
+function getPipelineVersionId(run?: ApiRun | ApiJob): string | null {
+  return run &&
+    run.resource_references &&
+    run.resource_references.some(
+      ref => ref.key && ref.key.type && ref.key.type === ApiResourceType.PIPELINEVERSION,
+    )
+    ? run.resource_references.find(
+        ref => ref.key && ref.key.type && ref.key.type === ApiResourceType.PIPELINEVERSION,
```
Comment (Bobgy) [inline frontend/src/lib/RunUtils.ts]: nit: move this to a util function

File: frontend/src/pages/PipelineDetails.tsx [hunk 5]
```diff
@@ -107,25 +112,41 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+        () => {
+          return this.state.selectedVersion
+            ? this.state.selectedVersion.id!
+            : pipelineVersionIdFromParams
+            ? pipelineVersionIdFromParams
+            : '';
+        },
+      )
+      .newPipelineVersion('Upload pipeline version', () =>
+        this.state.pipeline ? this.state.pipeline.id! : '',
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: Done

File: frontend/src/lib/Buttons.ts [hunk 6]
```diff
@@ -241,6 +273,23 @@ export default class Buttons {
     return this;
   }
 
+  public newRunFromPipelineVersion(
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/lib/Buttons.ts]: Done

File: frontend/src/lib/Buttons.ts [hunk 7]
```diff
@@ -304,7 +366,7 @@ export default class Buttons {
       outlined: true,
       style: { minWidth: 160 },
       title: 'Upload pipeline',
-      tooltip: 'Upload pipeline',
+      tooltip: 'Create a new pipeline and its first version',
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/lib/Buttons.ts]: Yes, this button leads to a page that can upload new pipeline or upload new pipeline version under existing pipeline.
Done

File: frontend/src/lib/Buttons.ts [hunk 8]
```diff
@@ -566,6 +646,24 @@ export default class Buttons {
     this._props.history.push(RoutePage.NEW_RUN + searchString);
   }
 
+  private _createNewRunFromPipelineVersion(pipelineId?: string, pipelineVersionId?: string): void {
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/lib/Buttons.ts]: Done

File: frontend/tslint.json (hunk 1 above)
Comment (jingzhang36) [reply frontend/tslint.json]: Done

File: frontend/src/pages/RunList.tsx (hunk 2 above)
Comment (jingzhang36) [reply frontend/src/pages/RunList.tsx]: done

File: frontend/src/pages/PipelineDetails.tsx (hunk 3 above)
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: done

File: frontend/src/lib/RunUtils.ts (hunk 4 above)
Comment (jingzhang36) [reply frontend/src/lib/RunUtils.ts]: Did you mean moving this to another util func to be called by this util func?

**Comment by jingzhang36** [comment]
> Test failed with
> 
> ```
> Checking formatting...
> src/lib/Buttons.ts
> src/lib/RunUtils.ts
> src/pages/RunList.test.tsx
> src/pages/RunList.tsx
> Code style issues found in the above file(s). Forgot to run Prettier?
> ```

I re-ran "npm run format". But still see formatting complaints...

**Comment by Bobgy** [comment]
Note setuptools is not a real failure, one integration test failed at the end of log.

File: frontend/src/lib/RunUtils.ts [hunk 9]
```diff
@@ -66,6 +66,32 @@ function getPipelineName(run?: ApiRun | ApiJob): string | null {
...
+      )!.key!.id!
```
Comment (Bobgy) [inline frontend/src/lib/RunUtils.ts]: Recommend being safe and not use !, provide a default value if key or id is null

File: frontend/src/pages/PipelineList.test.tsx [hunk 10]
```diff
@@ -467,142 +485,13 @@ describe('PipelineList', () => {
...
         'Failed to delete pipeline: test-pipeline-id0 with error: "woops, failed!"\n\n' +
         'Failed to delete pipeline: test-pipeline-id1 with error: "woops, failed!"',
-      title: 'Failed to delete 2 pipelines',
+      title: 'Failed to delete some pipelines and/or some pipeline versions',
     });
 
     // Should show snackbar for the one successful deletion
     expect(updateSnackbarSpy).toHaveBeenLastCalledWith({
-      message: 'Delete succeeded for 2 pipelines',
+      message: 'Deletion succeeded for 2 pipelines',
```
Comment (Bobgy) [inline frontend/src/pages/PipelineList.test.tsx]: Can you add a case for deleting a pipeline and a version together? Because that logic is very complex. I want to make sure it is covered.

File: frontend/src/pages/PipelineList.tsx [hunk 11]
```diff
@@ -145,18 +195,34 @@ class PipelineList extends Page<{}, PipelineListState> {
...
+  // (1) changes of selected pipeline ids, and will be stored in "this.state.selectedIds" or
+  // (2) changes of selected pipeline version ids, and will be stored in "selectedVersionIds" with key "pipelineId"
+  private _selectionChanged(pipelineId: string | undefined, selectedIds: string[]): void {
+    if (!!pipelineId) {
+      // Update selected pipeline version ids.
+      this.setStateSafe({
+        selectedVersionIds: { ...this.state.selectedVersionIds, ...{ [pipelineId!]: selectedIds } },
+      });
+      const actions = this.props.toolbarProps.actions;
+      actions[ButtonKeys.DELETE_RUN].disabled =
```
Comment (Bobgy) [inline frontend/src/pages/PipelineList.tsx]: This looks strange, why are we mutating props?

File: frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap [hunk 12]
```diff
@@ -64,6 +64,7 @@ exports[`PipelineDetails closes side panel when close button is clicked 1`] = `
             <div>
               test-pipeline-id
             </div>
+            0
```
Comment (Bobgy) [inline frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap]: I'm curious why this is 0?

File: frontend/src/pages/PipelineDetails.tsx [hunk 13]
```diff
@@ -372,47 +445,99 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+        actions[ButtonKeys.DELETE_RUN].disabled = true;
+        this.props.updateToolbar({ actions });
+      } else {
+        // Fetch manifest for the selected version under this pipeline.
+        pageTitle = pipeline.name!.concat(' (', selectedVersion!.name!, ')');
+        try {
+          // TODO(jingzhang36): pagination not proper here. so if many versions,
+          // the page size value should be?
+          versions =
+            (await Apis.pipelineServiceApi.listPipelineVersions('PIPELINE', pipelineId, 50))
```
Comment (Bobgy) [inline frontend/src/pages/PipelineDetails.tsx]: Just double check, does this return newest 50 versions? I think behavior is acceptable in this case.

File: frontend/src/pages/PipelineDetails.tsx [hunk 14]
```diff
@@ -150,12 +164,12 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
         )
         .delete(
           () =>
-            this.state.pipeline
-              ? [this.state.pipeline.id!]
-              : pipelineIdFromParams
-              ? [pipelineIdFromParams]
+            this.state.selectedVersion
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: Done

File: frontend/src/pages/PipelineDetails.tsx [hunk 15]
```diff
@@ -302,11 +372,14 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
     const fromRunId = new URLParser(this.props).get(QUERY_PARAMS.fromRunId);
 
     let pipeline: ApiPipeline | null = null;
+    let version: ApiPipelineVersion | null = null;
     let templateString = '';
-    let template: Workflow | undefined;
+    // let template: Workflow | undefined;
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: Done

File: frontend/src/pages/PipelineDetails.tsx [hunk 16]
```diff
@@ -289,6 +340,25 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
   }
 
+  public async handleVersionSelected(versionId: string): Promise<void> {
+    if (this.state.pipeline) {
+      const selectedVersion = (this.state.versions || []).find(v => v.id === versionId);
+      const selectedVersionPipelineTemplate = await this._getTemplateString(
+        this.state.pipeline.id!,
+        versionId,
+      );
+      // TODO(rjbauer): remove last history entry
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: Done

File: frontend/src/lib/Buttons.ts [hunk 17]
```diff
@@ -551,15 +616,18 @@ export default class Buttons {
...
 
     if (fromRunId) {
       searchString = this._urlParser.build(Object.assign({ [QUERY_PARAMS.fromRunId]: fromRunId }));
     } else {
       searchString = this._urlParser.build(
-        Object.assign({ [QUERY_PARAMS.pipelineId]: pipelineId || '' }),
+        Object.assign({
+          [QUERY_PARAMS.pipelineId]: pipelineId || '',
+          [QUERY_PARAMS.pipelineVersionId]: pipelineVersionId || '',
+        }),
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/lib/Buttons.ts]: done

File: frontend/src/lib/RunUtils.ts (hunk 9 above)
Comment (jingzhang36) [reply frontend/src/lib/RunUtils.ts]: Actually, the 3 !s have been checked by (1) some (2) ref.key (3) ref.key.id respectively. So the all !s just to take the value and otherwise, if any fails, it goes to null branch.

File: frontend/src/pages/PipelineDetails.tsx [hunk 18]
```diff
@@ -213,6 +235,35 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+                                  value={
+                                    selectedVersion
+                                      ? selectedVersion.id
+                                      : pipeline.default_version!.id!
+                                  }
+                                  onChange={event => this.handleVersionSelected(event.target.value)}
+                                  inputProps={{ id: 'version-selector', name: 'selectedVersion' }}
+                                >
+                                  {versions.map((v, i) => (
+                                    <MenuItem key={i} value={v.id}>
```
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: done

File: frontend/src/pages/PipelineList.tsx (hunk 11 above)
Comment (jingzhang36) [reply frontend/src/pages/PipelineList.tsx]: My understanding is when no checkbox is checked (no pipeline or no version is selected), the delete button is gray. I kept the current behavior, and whether to change this behavior or not is open to discussion.

File: frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap (hunk 12 above)
Comment (jingzhang36) [reply frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap]: My guess is: this place is the version selector; and we show default version (index 0) when we open pipeline details by clicking the pipeline (instead of clicking a particular pipeline version)

File: frontend/src/pages/PipelineDetails.tsx (hunk 13 above)
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: I believe the backend default is asc, if nothing specified. So oldest first. You prefer desc? I changed it.

File: frontend/src/pages/PipelineDetails.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: An issue after I changed it. The test complains about selectedIds.map is not a func. So I tried
(pipelineVersionIdFromParams) as string[] || [] and (pipelineVersionIdFromParams || []) as string[]. But doesn't solve it.

File: frontend/src/pages/PipelineDetails.tsx (hunk 13 above)
Comment (Bobgy) [reply frontend/src/pages/PipelineDetails.tsx]: Newest sounds good. I believe usually people check latest and near-latest versions

File: frontend/src/lib/RunUtils.ts (hunk 9 above)
Comment (Bobgy) [reply frontend/src/lib/RunUtils.ts]: SG, let me introduce optional chaining, so it won't be so painful later

File: frontend/src/pages/PipelineDetails.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (jingzhang36) [reply frontend/src/pages/PipelineDetails.tsx]: Done

File: frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap (hunk 12 above)
Comment (jingzhang36) [reply frontend/src/pages/__snapshots__/PipelineDetails.test.tsx.snap]: done
